sympy/ntheory/factor_.py

```diff
@@ -608,6 +611,117 @@ def _factors():
+        def get_external(self, n: int) -> list[int] | None:
+            res = requests.get("http://factordb.com/api", params={"query": str(n)})
+            if res.status_code != requests.codes.ok:
+                return None
+            j = res.json()
+            if j.get("status") in ["FF", "P"]:
+                return list(int(p) for p, _ in j.get("factors"))
+
+        factor_cache.get_external = get_external
+
+    Be aware that writing this code will trigger internet access
+    to factordb.com when calling ``factorint``.
+
+    """
+    def __init__(self, maxsize: int | None = None):
+        self._cache: OrderedDict[int, int] = OrderedDict()
+        self.maxsize = maxsize
+
+    def __len__(self) -> int:
+        return len(self._cache)
+
+    def __contains__(self, n) -> bool:
+        return n in self._cache
+
+    def __getitem__(self, n: int) -> int:
+        factor = self.get(n)
+        if factor is None:
+            raise KeyError(f"{n} does not exist.")
+        return factor
+
+    def __setitem__(self, n: int, factor: int):
+        if not (1 < factor <= n and n % factor == 0 and isprime(factor)):
+            raise ValueError(f"{factor} is not a prime factor of {n}")
+        self._cache[n] = max(self._cache.get(n, 0), factor)
+        if self.maxsize is not None and len(self._cache) > self.maxsize:
+            self._cache.popitem(False)
+
+    def __delitem__(self, n: int):
+        if n not in self._cache:
+            raise KeyError(f"{n} does not exist.")
+        del self._cache[n]
+
+    def __iter__(self):
+        return self._cache.__iter__()
+
+    def cache_clear(self) -> None:
+        """ Clear the cache """
+        self._cache = OrderedDict()
+
+    @property
+    def maxsize(self) -> int | None:
+        """ Returns the maximum cache size; if ``None``, it is unlimited. """
+        return self._maxsize
+
+    @maxsize.setter
+    def maxsize(self, value: int | None) -> None:
+        if value is not None and value <= 0:
+            raise ValueError("maxsize must be None or a non-negative integer.")
+        self._maxsize = value
+        if value is not None:
+            while len(self._cache) > value:
+                self._cache.popitem(False)
+
+    def get(self, n: int, default=None):
+        """ Return the prime factor of ``n``.
+        If it does not exist in the cache, return the value of ``default``.
+        """
+        if n <= sieve._list[-1]:
+            if sieve._list[bisect_left(sieve._list, n)] == n:
+                return n
+        if n in self._cache:
+            self._cache.move_to_end(n)
+            return self._cache[n]
+        if factors := self.get_external(n):
+            self.add(n, factors)
+            return self._cache[n]
+        return default
+
+    def add(self, n: int, factors: list[int]) -> None:
+        for p in sorted(factors, reverse=True):
+            self[n] = p
+            n, _ = remove(n, p)
+
+    def get_external(self, n: int) -> list[int] | None:
```

Comment:
I notice that this function is not implemented before merge

Reply:
The `get_external` function is meant to be implemented by the user and is not intended to be provided by SymPy. If the user does not implement it, the function will always return `None`, and lines 709 and 710 will never be executed.

Does this explanation work for you? I was not entirely sure about the intention behind "before merge."

Reply:
No, I don't think that it is possible for user to override get_external to be working.
I think that it is only possible with overriding, but sympy uses an instance of FactorCache so there is no option to override get_external for that function.

Maybe it is possible with factor_cache.get_external = ... but the type annotation could be different

Reply:
Okay, I notice that it is documented

Reply:
But anyway, having a in-place replacement of method with function with `self` parameter looked quite confusing to me